lib/kcerts/ssh.go

```diff
@@ -0,0 +1,116 @@
+package kcerts
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string) error {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return err
+	}
+	hDir = hDir + "/"
+	if _, err := os.Stat(hDir + SSHDir); os.IsNotExist(err) {
+		return fmt.Errorf("ssh directory %s does not exist, please create it", hDir+SSHDir)
+	}
+	qualifiedKnownHosts := hDir + SSHDir + "/" + KnownHosts
+	if _, err := os.Stat(qualifiedKnownHosts); os.IsNotExist(err) {
+		return fmt.Errorf("ssh authorized hosts file %s does not exist, please create it", qualifiedKnownHosts)
+	}
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	existingKnownHostsContent, err := ioutil.ReadFile(qualifiedKnownHosts)
```

Comment:
Regarding both the os.Stat and ReadFile 

I disagree with this because there isn't a valuable reason to ignore the errors. If we ignore the errors and don't os.Stat: The linter complains, it provides less info for the end user, and can possibly obfuscate errors. Generally I like to provide as much stdout info as possible especially when mucking about files users routinely touch. 

That being said, it should try and make them if it doesn't exist, so I implemented that with proper errors.

Reply:
> I disagree with this because there isn't a valuable reason to ignore the errors.
> If we ignore the errors and don't os.Stat: The linter complains, it provides less info for the end user, and can possibly obfuscate errors. Generally I like to provide as much stdout info as possible especially when mucking about files users routinely touch.

I don't disagree with the goals, but I think the code here now is significantly more complex than it could be, and is actually making the errors harder.

Doing Stat() and then the operation (eg, Read...(), Open...(), Write...()) is common in some languages, but should generally be frowned upon as an anti-pattern: Stat() can fail (or succeed) for reasons different from that of the operation, and introduces race conditions (in between Stat() and Open(), for example, the file may have been removed). Instead of one possible error, you now have two, and both need to be managed (as, for example, a Stat() succeeding and indicating a file exists and is readable does not imply the next Open() will also succeed and allow you to read the file - fs acls, syscall filtering, ...).

Instead of using Stat() to check if a file can be written or read, just try to write or read into the file, and *check the returned error code*.

In the suggestion I had made above, I just wanted your code to work correctly in the absence of the known hosts file (perfectly legal), absence of directory, empty file, ...